File: app/src/main/java/com/kota/Telnet/TelnetOutputBuilder.java
package com.kota.Telnet;

import com.kota.ASFramework.Thread.ASRunner;
import com.kota.DataPool.MutableByteBuffer;
import com.kota.Telnet.Reference.TelnetDefs;
import com.kota.Telnet.Reference.TelnetKeyboard;
import com.kota.TextEncoder.U2BEncoder;

public class TelnetOutputBuilder {
    private MutableByteBuffer _buffer = MutableByteBuffer.createMutableByteBuffer();

    public static TelnetOutputBuilder create() {
        return new TelnetOutputBuilder();
    }

    public void sendToServerInBackground() {
        sendToServerInBackground(0);
    }

    public void sendToServerInBackground(int channel) {
        TelnetClient.getClient().sendDataToServerInBackground(build(), channel);
    }

    public void sendToServer() {
        if (ASRunner.isMainThread()) {
            sendToServerInBackground(0);
        } else {
            sendToServer(0);
        }
    }

    private void sendToServer(int channel) {
        TelnetClient.getClient().sendDataToServer(build(), channel);
    }

    public byte[] build() {
        this._buffer.close();
        byte[] data = this._buffer.toByteArray();
        MutableByteBuffer.recycleMutableByteBuffer(this._buffer);
        return data;
    }

    public TelnetOutputBuilder pushData(byte data) {
        this._buffer.put(data);
        return this;
    }

    public void pushData(byte[] data) {
        for (byte pushData : data) {
            pushData(pushData);
        }
    }

    public TelnetOutputBuilder pushKey(int key) {
        pushData(TelnetKeyboard.getKeyData(key));
        return this;
    }

    public TelnetOutputBuilder pushString(String str) {
        try {
            pushData(U2BEncoder.getInstance().encodeToBytes(str.getBytes(TelnetDefs.CHARSET), 0));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return this;
    }
}

File: app/src/main/java/com/kota/telnet/TelnetOutputBuilder.kt
package com.kota.telnet

import android.util.Log
import com.kota.asFramework.thread.ASRunner
import com.kota.dataPool.MutableByteBuffer
import com.kota.telnet.reference.TelnetDef
import com.kota.telnet.reference.TelnetKeyboard.getKeyData
import com.kota.textEncoder.U2BEncoder

class TelnetOutputBuilder {
    private val byteBuffers: MutableByteBuffer = MutableByteBuffer.createMutableByteBuffer()

    @JvmOverloads
    fun sendToServerInBackground(channel: Int = 0) {
        TelnetClient.Companion.client?.sendDataToServerInBackground(build(), channel)
    }

    fun sendToServer() {
        if (ASRunner.isMainThread) {
            sendToServerInBackground(0)
        } else {
            sendToServer(0)
        }
    }

    private fun sendToServer(channel: Int) {
        TelnetClient.Companion.client?.sendDataToServer(build(), channel)
    }

    fun build(): ByteArray? {
        this.byteBuffers.close()
        val data = this.byteBuffers.toByteArray()
        MutableByteBuffer.recycleMutableByteBuffer(this.byteBuffers)
        return data
    }

    fun pushData(data: Byte): TelnetOutputBuilder {
        this.byteBuffers.put(data)
        return this
    }

    fun pushData(data: ByteArray) {
        for (pushData in data) {
            pushData(pushData)
        }
    }

    fun pushKey(key: Int): TelnetOutputBuilder {
        pushData(getKeyData(key))
        return this
    }

    fun pushString(str: String): TelnetOutputBuilder {
        try {
            pushData(
                U2BEncoder.instance!!.encodeToBytes(str.toByteArray(charset(TelnetDef.CHARSET)), 0)
            )
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
        }
        return this
    }

    companion object {
        @JvmStatic
        fun create(): TelnetOutputBuilder {
            return TelnetOutputBuilder()
        }
    }
}
